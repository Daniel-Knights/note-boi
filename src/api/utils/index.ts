export * from './decorators';
export * from './error';
export * from './http';
