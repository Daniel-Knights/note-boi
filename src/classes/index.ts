export * from './appError';
export * from './debounceQueue';
export * from './encryptor';
export * from './fetchBuilder';
export * from './keyStore';
export * from './storage';
export * from './unsyncedNotesManager';
