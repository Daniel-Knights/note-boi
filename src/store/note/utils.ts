import { Note } from '../../classes';
import { isEmptyNote } from '../../utils';

import { deleteNote } from './operations';
import { noteState } from './state';

/** Sorts notes in descending order by timestamp. */
export const sortNotesFn = (a: Note, b: Note): number => b.timestamp - a.timestamp;

/** Sorts {@link noteState.notes} in descending order by timestamp. */
export const sortStateNotes = (): Note[] => noteState.notes.sort(sortNotesFn);

/** Finds note index within {@link noteState.notes}. */
export function findNoteIndex(uuid?: string): number {
  return noteState.notes.findIndex((nt) => nt.uuid === uuid);
}

/** Finds note within {@link noteState.notes}. */
export function findNote(uuid?: string): Note | undefined {
  return noteState.notes.find((nt) => nt.uuid === uuid);
}

/**
 * Returns true if note is either {@link noteState.selectedNote}
 * or within {@link noteState.extraSelectedNotes}.
 */
export function isSelectedNote(note: Note): boolean {
  return (
    note.uuid === noteState.selectedNote.uuid ||
    noteState.extraSelectedNotes.some((nt) => nt?.uuid === note.uuid)
  );
}

/** Deletes {@link noteState.selectedNote} when note is empty. */
export function clearEmptyNote(): void {
  const isValidClear = noteState.notes.length > 1;
  if (!isValidClear) return;

  const foundNote = findNote(noteState.selectedNote.uuid);
  if (!foundNote) return;

  if (isEmptyNote(foundNote)) {
    deleteNote(noteState.selectedNote.uuid);
  }
}
