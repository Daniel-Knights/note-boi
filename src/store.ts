import { reactive } from 'vue';
import { invoke } from '@tauri-apps/api/tauri';
import { v4 as uuidv4 } from 'uuid';

import { isEmptyNote } from './utils';

export class Note {
  readonly id = uuidv4();
  body = {
    delta: '',
    text: '',
  };
  timestamp = Date.now();
}

interface State {
  notes: Note[];
  selectedNote: Note;
}

export const state = reactive<State>({
  notes: [],
  selectedNote: new Note(),
});

/** Sorts notes in descending order by timestamp. */
function sortNotes() {
  state.notes.sort((a, b) => b.timestamp - a.timestamp);
}

/** Finds note index within {@link state.notes}. */
function findNoteIndex(id: string) {
  return state.notes.findIndex((nt) => nt.id === id);
}

/** Finds note within {@link state.notes}. */
export function findNote(id: string): Note | undefined {
  return state.notes.find((nt) => nt.id === id);
}

/** Deletes {@link state.selectedNote} when both `title` and `body` are empty. */
function clearEmptyNote(): void {
  const isValidClear = state.notes.length > 1;

  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote) return;

  if (isValidClear && isEmptyNote(foundNote)) {
    deleteNote(state.selectedNote.id);
  }
}

const selectEvent = new CustomEvent('note-select');

/**
 * Looks for note with given `id` in {@link state.notes},
 * and sets it to {@link state.selectedNote}.
 */
export function selectNote(id: string): void {
  if (state.selectedNote.id === id) return;

  clearEmptyNote();

  const foundNote = findNote(id);
  if (foundNote) state.selectedNote = { ...foundNote };

  document.dispatchEvent(selectEvent);
}

/** Fetches all notes and updates {@link state}. */
export async function getAllNotes(): Promise<void> {
  const fetchedNotes = await invoke<Note[]>('get_all_notes').catch(console.error);

  const hasNotes = fetchedNotes && fetchedNotes.length > 0;
  if (!hasNotes) return newNote();

  state.notes = fetchedNotes;

  sortNotes();

  if (isEmptyNote(fetchedNotes[0])) {
    state.notes[0].timestamp = Date.now();
  }

  state.selectedNote = { ...state.notes[0] };

  document.dispatchEvent(selectEvent);
}

/** Deletes note with the given `id`. */
export function deleteNote(id: string): void {
  state.notes.splice(findNoteIndex(id), 1);
  state.selectedNote = { ...state.notes[0] };

  sortNotes();

  invoke('delete_note', { id }).catch(console.error);
}

/** Creates an empty note. */
export function newNote(): void {
  const foundNote = findNote(state.selectedNote.id);

  // Only update timestamp if selected note is empty
  if (foundNote && isEmptyNote(foundNote)) {
    state.selectedNote.timestamp = Date.now();
    return;
  }

  const freshNote = new Note();

  state.notes.unshift(freshNote);
  state.selectedNote = { ...freshNote };

  invoke('new_note', { note: { ...freshNote } }).catch(console.error);

  document.dispatchEvent(selectEvent);
}

/** Edits note body on Quill `text-change`. */
export function editBody(delta: string, text: string): void {
  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote || text === foundNote.body.text) return;

  const timestamp = Date.now();
  foundNote.timestamp = timestamp;
  state.selectedNote.timestamp = timestamp;

  foundNote.body = { delta, text };

  sortNotes();

  invoke('edit_note', { note: { ...foundNote } }).catch(console.error);
}
