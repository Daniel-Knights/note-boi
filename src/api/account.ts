import * as dialog from '@tauri-apps/plugin-dialog';

import {
  AppError,
  Encryptor,
  ERROR_CODE,
  ErrorConfig,
  FetchBuilder,
  KeyStore,
} from '../classes';
import { noteState } from '../store/note';
import { resetAppError, syncState } from '../store/sync';
import { tauriInvoke } from '../utils';

import { clientSideLogout } from './auth';
import {
  parseErrorRes,
  resIsOk,
  route,
  throwEncryptorError,
  throwFetchError,
} from './utils';

export const changePassword = route(async (): Promise<void> => {
  const errorConfig = {
    code: ERROR_CODE.CHANGE_PASSWORD,
    retry: { fn: changePassword },
    display: {
      form: true,
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof changePassword>, 'message'>;

  const newKey = await Encryptor.generatePasswordKey(syncState.newPassword);

  await KeyStore.storeKey(newKey);

  const [accessToken, encryptedNotes] = await Promise.all([
    tauriInvoke('get_access_token', {
      username: syncState.username,
    }),
    Encryptor.encryptNotes(noteState.notes, newKey).catch((err) =>
      throwEncryptorError(errorConfig, err)
    ),
  ]);
  if (!accessToken || !encryptedNotes) return;

  const res = await new FetchBuilder('/account/change-password')
    .method('PUT')
    .withAuth(syncState.username, accessToken)
    .body({
      current_password: syncState.password,
      new_password: syncState.newPassword,
      notes: encryptedNotes,
    })
    .fetch(syncState.username)
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    resetAppError();

    syncState.password = '';
    syncState.newPassword = '';
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});

export const deleteAccount = route(async (): Promise<void> => {
  const askRes = await dialog.ask('Are you sure?', {
    title: 'Delete account',
    kind: 'warning',
  });
  if (!askRes) return;

  const errorConfig = {
    code: ERROR_CODE.DELETE_ACCOUNT,
    retry: { fn: deleteAccount },
    display: {
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof deleteAccount>, 'message'>;

  const accessToken = await tauriInvoke('get_access_token', {
    username: syncState.username,
  });

  const res = await new FetchBuilder('/account/delete')
    .method('DELETE')
    .withAuth(syncState.username, accessToken)
    .fetch()
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    await clientSideLogout();
    resetAppError();
    syncState.unsyncedNotes.clear(true);
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});
