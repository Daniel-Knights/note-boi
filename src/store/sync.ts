import { reactive } from 'vue';
import { http, invoke } from '@tauri-apps/api';
import type { Response } from '@tauri-apps/api/http';

import { getAllNotes, Note, state as noteState } from './note';
import { isDev, tauriEmit } from '../utils';

export enum ErrorType {
  None,
  Auth,
  Push,
  Pull,
  Logout,
}

interface State {
  username: string;
  password: string;
  token: string;
  hasUnsyncedNotes: boolean;
  isLoading: boolean;
  isLogin: boolean; // For switching login/signup form
  autoSyncEnabled: boolean;
  error: {
    type: ErrorType;
    message: string;
  };
}

const autoSync = localStorage.getItem('auto-sync') as 'true' | 'false' | null;

export const state = reactive<State>({
  username: localStorage.getItem('username') || '',
  password: '',
  token: localStorage.getItem('token') || '',
  hasUnsyncedNotes: false,
  isLoading: false,
  isLogin: true,
  autoSyncEnabled: autoSync !== null ? autoSync === 'true' : true,
  error: {
    type: ErrorType.None,
    message: '',
  },
});

// Utils //

/** Parses an error response and returns a formatted message. */
function parseErrorRes(res: Response<Record<string, unknown>>) {
  const unknownErrorMessage = 'Unknown error, please try again';

  if (!res.data) return unknownErrorMessage;

  return typeof res.data.error === 'string' ? res.data.error : unknownErrorMessage;
}

/** Resets {@link state.error}. */
export function resetError(): void {
  state.error = { type: ErrorType.None, message: '' };
}

/** Wrapper for {@link http.fetch}. */
function tauriFetch<T>(
  endpoint: string,
  method: 'POST' | 'PUT',
  payload?: Record<string, unknown>
): Promise<Response<T>> {
  const baseUrl = isDev() ? 'http://localhost:8000' : 'TODO';

  return http.fetch<T>(`${baseUrl}/api${endpoint}`, {
    method,
    body: {
      type: 'Json',
      payload,
    },
  });
}

function clientSideLogout() {
  state.token = '';
  state.username = '';

  localStorage.removeItem('token');
  localStorage.removeItem('username');
}

/** Sets user preference for auto-syncing to {@link state} and `localStorage` */
export function setAutoSync(enabled: boolean): void {
  state.autoSyncEnabled = enabled;
  localStorage.setItem('auto-sync', enabled ? 'true' : 'false');
}

// Routes //

export async function login(): Promise<void> {
  state.isLoading = true;

  const res = await tauriFetch<Record<string, string | Note[]>>('/login', 'POST', {
    username: state.username,
    password: state.password,
  });

  if (res.ok) {
    resetError();
    tauriEmit('login');

    await invoke('sync_all_local_notes', { notes: res.data.notes }).catch(console.error);
    await getAllNotes().catch(console.error);

    state.token = res.data.token as string;
    state.password = '';

    localStorage.setItem('username', state.username);
    localStorage.setItem('token', state.token);
  }

  state.isLoading = false;

  if (!res.ok) {
    state.error = {
      type: ErrorType.Auth,
      message: parseErrorRes(res),
    };

    console.error(res.data);
  }
}

export async function signup(): Promise<void> {
  state.isLoading = true;

  const res = await tauriFetch<Record<string, string>>('/signup', 'POST', {
    username: state.username,
    password: state.password,
    notes: noteState.notes,
  });

  state.isLoading = false;

  if (res.ok) {
    resetError();
    tauriEmit('login');

    state.token = res.data.token;
    state.password = '';

    localStorage.setItem('username', state.username);
    localStorage.setItem('token', state.token);
  } else {
    state.error = {
      type: ErrorType.Auth,
      message: parseErrorRes(res),
    };

    console.error(res.data);
  }
}

export async function logout(): Promise<void> {
  if (!state.token) return; // Prevent bug where event.emit triggers event.listen

  state.isLoading = true;

  const res = await tauriFetch<Record<string, never | string>>('/logout', 'POST', {
    username: state.username,
    token: state.token,
  });

  state.isLoading = false;

  if (res.ok) {
    resetError();
    tauriEmit('logout');
    clientSideLogout();
  } else {
    state.error = {
      type: ErrorType.Logout,
      message: parseErrorRes(res),
    };
  }
}

export async function pull(): Promise<void> {
  state.isLoading = true;

  const res = await tauriFetch<Record<string, string | Note[]>>('/notes', 'POST', {
    username: state.username,
    token: state.token,
  });

  if (res.ok) {
    resetError();
    await invoke('sync_all_local_notes', { notes: res.data.notes }).catch(console.error);
    await getAllNotes().catch(console.error);
  }

  state.isLoading = false;

  if (!res.ok) {
    state.error = {
      type: ErrorType.Pull,
      message: parseErrorRes(res),
    };

    // User not found
    if (res.status === 404) clientSideLogout();

    console.error(res.data);
  }
}

export async function push(): Promise<void> {
  state.isLoading = true;

  const res = await tauriFetch<Record<string, never | string>>('/notes', 'PUT', {
    username: state.username,
    token: state.token,
    notes: noteState.notes,
  });

  state.isLoading = false;

  if (res.ok) {
    resetError();
    state.hasUnsyncedNotes = false;
  } else {
    state.error = {
      type: ErrorType.Push,
      message: parseErrorRes(res),
    };

    console.error(res.data);
  }
}

document.addEventListener('note-unsynced', () => {
  if (state.token) {
    state.hasUnsyncedNotes = true;
  }
});
