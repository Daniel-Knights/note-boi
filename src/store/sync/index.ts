import { http } from '@tauri-apps/api';
import type { Response } from '@tauri-apps/api/http';
import { reactive } from 'vue';

import { STORAGE_KEYS } from '../../constant';
import { isDev } from '../../utils';
import { UnsyncedNoteIds, unsyncedNoteIds } from './note';

export enum ErrorType {
  None,
  Auth,
  Push,
  Pull,
  Logout,
}

export const syncState = reactive({
  username: localStorage.getItem(STORAGE_KEYS.USERNAME) || '',
  password: '',
  token: localStorage.getItem(STORAGE_KEYS.TOKEN) || '',
  isLoading: false,
  isLogin: true, // For switching login/signup form
  error: {
    type: ErrorType.None,
    message: '',
  },
  unsyncedNoteIds: <UnsyncedNoteIds>{
    new: unsyncedNoteIds.new || '',
    edited: new Set<string>(unsyncedNoteIds.edited),
    deleted: new Set<string>(unsyncedNoteIds.deleted),
    get size() {
      return this.edited.size + this.deleted.size;
    },
    clear() {
      this.new = '';
      this.edited.clear();
      this.deleted.clear();
      localStorage.removeItem(STORAGE_KEYS.UNSYNCED);
    },
    add(ids): void {
      if (ids.new !== undefined) this.new = ids.new;

      ids.edited?.forEach((id) => this.edited.add(id));

      // Only track deleted notes if logged in
      ids.deleted?.forEach((id) => {
        this.deleted.add(id);

        if (this.edited.has(id)) {
          this.edited.delete(id);
        }
      });

      if (this.edited.has(this.new) || this.deleted.has(this.new)) {
        this.new = '';
      }

      localStorage.setItem(
        STORAGE_KEYS.UNSYNCED,
        JSON.stringify({
          new: this.new,
          edited: [...this.edited],
          deleted: [...this.deleted],
        })
      );
    },
  },
});

// Utils //

/** Parses an error response and returns a formatted message. */
export function parseErrorRes(res: Response<Record<string, unknown>>): string {
  const unknownErrorMessage = 'Unknown error, please try again';

  if (!res.data) return unknownErrorMessage;

  return typeof res.data.error === 'string' ? res.data.error : unknownErrorMessage;
}

/** Resets {@link state.error}. */
export function resetError(): void {
  syncState.error = { type: ErrorType.None, message: '' };
}

/** Wrapper for {@link http.fetch}. */
export function tauriFetch<T>(
  endpoint: string,
  method: 'POST' | 'PUT',
  payload?: Record<string, unknown>
): Promise<Response<T>> {
  const baseUrl = isDev()
    ? 'http://localhost:8000'
    : 'https://note-boi-server.herokuapp.com';

  return http.fetch<T>(`${baseUrl}/api${endpoint}`, {
    method,
    body: {
      type: 'Json',
      payload,
    },
  });
}

/** Catches hanging requests (e.g. due to server error). */
export function catchHang(err: unknown, type: ErrorType): void {
  console.error(err);

  syncState.isLoading = false;
  syncState.error = {
    type,
    message: 'Request failed',
  };
}

export { autoPush, push, pull, syncNotes, UnsyncedNoteIds } from './note';
export { clientSideLogout, login, signup, logout } from './auth';
export { deleteAccount } from './account';
