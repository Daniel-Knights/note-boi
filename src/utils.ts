import { event, invoke } from '@tauri-apps/api';
import { EventCallback, UnlistenFn } from '@tauri-apps/api/event';

import { TauriCommand } from './constant';
import type { Note } from './store/note';

/** `process.env.NODE_ENV === 'development'`. */
export function isDev(): boolean {
  return process.env.NODE_ENV === 'development';
}

/** Formats Unix time to date-time. */
export function unixToDateTime(unixTime: number): string {
  return Intl.DateTimeFormat([], {
    dateStyle: 'medium',
    timeStyle: isDev() ? 'long' : 'short',
  }).format(unixTime);
}

/** Gets key value from `localStorage` and parses it as JSON. */
export function localStorageParse(key: string): ReturnType<typeof JSON.parse> {
  return JSON.parse(localStorage.getItem(key) || '{}');
}

/** Returns `true` if string consists of only whitespace characters or is empty. */
export function isWhitespaceOnly(text?: string): boolean {
  return text?.trim() === '';
}

/** Returns true if note content is empty or note is `undefined`. */
export function isEmptyNote(note?: Note): boolean {
  if (!note) return true;

  return isWhitespaceOnly(note.content.title) && isWhitespaceOnly(note.content.body);
}

/** Calls {@link event.emit}, with stronger typing for `id`. */
export function tauriEmit<T>(id: 'login' | 'logout', payload?: T): void {
  event.emit(id, payload);
}

/** Calls {@link event.listen}, with stronger typing for `id`. */
export function tauriListen(
  id:
    | 'reload'
    | 'new-note'
    | 'delete-note'
    | 'push-notes'
    | 'login'
    | 'logout'
    | 'signup'
    | `tauri://${string}`,
  cb: EventCallback<unknown>
): Promise<UnlistenFn> {
  return event.listen(id, cb);
}

/** Calls {@link invoke}, with stronger typing for `cmd`. */
export function tauriInvoke<T>(
  cmd: TauriCommand,
  args?: Parameters<typeof invoke>[1]
): Promise<T> {
  return invoke(cmd, args);
}
