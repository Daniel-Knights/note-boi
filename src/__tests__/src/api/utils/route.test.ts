import * as s from '../../../../store/sync';
import { route } from '../../../../api/utils';
import { ERROR_CODE } from '../../../../classes';
import { assertAppError, assertLoadingState } from '../../../utils';

describe('route', () => {
  it('Handles unknown errors', async () => {
    const err = new Error('unexpected');
    const consoleErrorSpy = vi.spyOn(console, 'error');
    const routeCb = vi.fn().mockRejectedValueOnce(err);
    const wrapped = route(routeCb);

    await assertLoadingState(async () => {
      const result = await wrapped('bar');

      expect(routeCb).toHaveBeenCalledWith('bar');
      expect(consoleErrorSpy).toHaveBeenCalledWith(`ERROR_CODE: ${ERROR_CODE.UNKNOWN}`);
      expect(consoleErrorSpy).toHaveBeenCalledWith('Original error:');
      expect(consoleErrorSpy).toHaveBeenCalledWith(err);

      assertAppError({
        code: ERROR_CODE.UNKNOWN,
        originalError: err,
        retry: { args: ['bar'] },
        display: { form: true, sync: true },
      });

      assert.isUndefined(result);
    });

    vi.clearAllMocks();

    await assertLoadingState(async () => {
      await s.syncState.appError.retry();

      expect(routeCb).toHaveBeenCalledWith('bar');
      expect(consoleErrorSpy).not.toHaveBeenCalled();
    });
  });
});
