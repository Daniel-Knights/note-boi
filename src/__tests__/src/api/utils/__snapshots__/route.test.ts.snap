// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`route > Handles unknown errors > KeyStore.getKey() 1`] = `undefined`;

exports[`route > Handles unknown errors > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`route > Handles unknown errors > Storage.get('THEME') 1`] = `null`;

exports[`route > Handles unknown errors > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`route > Handles unknown errors > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`route > Handles unknown errors > Storage.get('USERNAME') 1`] = `null`;

exports[`route > Handles unknown errors > allCalls 1`] = `[]`;

exports[`route > Handles unknown errors > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`route > Handles unknown errors > mockDb 1`] = `
{
  "deletedNoteIds": Set {},
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`route > Handles unknown errors > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`route > Handles unknown errors > openedPopup 1`] = `undefined`;

exports[`route > Handles unknown errors > selectedTheme 1`] = `"System"`;

exports[`route > Handles unknown errors > syncState 1`] = `
{
  "appError": AppError {
    "code": 1,
    "display": {
      "form": true,
      "sync": true,
    },
    "message": "An unknown error occurred",
    "originalError": [Error: unexpected],
    "retry": [Function],
    "retryConfig": {
      "args": [
        "bar",
      ],
      "fn": [Function],
    },
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": UnsyncedNotesManager {
    "deleted": Set {},
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`route > Handles unknown errors > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
